### ISSUE TITLE
Default bf16 (X6) dot algorithm gives wrong inverse SH transforms for single-level fields on GPU (jax ≤ 0.11.1), causing mass drift in hybrid SL runs

### ISSUE THREAD

**Issue by duncanwp** [neuralgcm/dinosaur#147]
On GPU with jax/jaxlib ≤ 0.11.1, the default float32 dot algorithm introduced in 1.5.0 (`BF16_BF16_F32_X6`) gives wrong results for the inverse spherical-harmonic transform of **single-level** fields, i.e. modal arrays with a leading dimension of size 1 such as `log_surface_pressure`. The error is ~2e-5 relative (≈200× the float32 floor) and is a fixed pattern every step. With `SemiLagrangianPrimitiveEquationsHybrid` this integrates into a large hemispheric mass drift: a Held–Suarez aquaplanet at T63L47 develops a −116 hPa NH−SH surface-pressure asymmetry within 30 days, against +0.03 hPa with `Precision.HIGHEST`. Sigma-coordinate runs weren't affected in our tests.

### Minimal reproduction (A100, jax 0.10.2 / 0.11.1)

```python
import functools
import numpy as np, jax, jax.numpy as jnp
from dinosaur import spherical_harmonic as sh

grid = sh.Grid.T42()
x = np.random.default_rng(0).standard_normal((2,) + grid.modal_shape).astype(np.float32)

def to_nodal(v, precision=None):
  if precision is not None:
    sh.einsum = functools.partial(jnp.einsum, precision=precision)
  return np.asarray(jax.jit(grid.to_nodal)(v), np.float64)

default = [to_nodal(x[:1]), to_nodal(x)[:1]]  # (1, M, L) vs (2, M, L)[:1]
exact = to_nodal(x[:1], jax.lax.Precision.HIGHEST)
for name, out in zip(['batch of 1', 'batch of 2'], default):
  print(f'{name}: max rel error {np.abs(out - exact).max() / np.abs(exact).max():.1e}')
```
```
batch of 1: max rel error 2.4e-05   # jax 0.10.2 and 0.11.1
batch of 2: max rel error 3.7e-07
batch of 1: max rel error 3.7e-07   # jax 0.11.2
```

### Cause: an XLA:GPU miscompile, fixed in jax 0.11.2

The bug is triggered only by the combination of a **compile-time-constant** operand (the numpy Legendre basis) and a **size-1** free dimension on the other operand, in the `'mjl,...ml->...mj'` inverse-Legendre contraction. Pass the basis as a jit argument, or use a batch ≥ 2, and the result is correct. JAX emits identical StableHLO in both cases, and disabling XLA's `constant_folding` or `algsimp` passes also fixes it. XLA's dot-strength-reduction rewrites the dot as multiply+reduce with a bitmask bf16 split. Algsimp then cancels the `dot_decomposer` transposes of the constant, leaving a layout-changing `bitcast-convert` that constant folding evaluates as a raw buffer reinterpretation. The constant's hi/mid/lo pieces come out permuted, so the split loses ~8 bits; an exact numpy emulation of this reproduces XLA's output to 1e-7. Present in jax 0.10.2, 0.11.0 and 0.11.1; gone in 0.11.2 (XLA `dcf304bc…` → `91888df6…`). We found no existing jax or XLA report.

Pure-JAX version:

```python
import numpy as np, jax, jax.numpy as jnp
P = np.random.default_rng(0).standard_normal((127, 96, 65)).astype(np.float32)
y = np.random.default_rng(1).standard_normal((1, 127, 65)).astype(np.float32)
ref = np.einsum('mjl,bml->bmj', P.astype(np.float64), y.astype(np.float64))
X6 = jax.lax.DotAlgorithmPreset.BF16_BF16_F32_X6
const = jax.jit(lambda v: jnp.einsum('mjl,bml->bmj', P, v, precision=X6))(y)
arg = jax.jit(lambda p, v: jnp.einsum('mjl,bml->bmj', p, v, precision=X6))(P, y)
for o in (const, arg):
  print(np.abs(np.asarray(o, np.float64) - ref).max() / np.abs(ref).max())
# jax <= 0.11.1: 2.8e-05, 1.1e-07   jax 0.11.2: 1.1e-07, 1.1e-07
```

### Possible mitigations in dinosaur
- Fall back to `Precision.HIGHEST` for float32 dots on jaxlib < 0.11.2, or
- avoid the triggering shape in the SH transforms (e.g. fold a size-1 leading dim away before the einsum).

In the meantime, downstream users can set `dinosaur.jax_numpy_utils.FLOAT32_DOT_ALGORITHM = jax.lax.Precision.HIGHEST` before building a model. Note that `spherical_harmonic.FAST_TRANSFORM_DOT_ALGORITHM` only reaches `FastSphericalHarmonics`; `RealSphericalHarmonics` resolves through `FLOAT32_DOT_ALGORITHM`.

**Comment by duncanwp** [comment]
A related but separate observation on the SPMD path: `FastSphericalHarmonics` uses `FAST_TRANSFORM_DOT_ALGORITHM` (`BF16_BF16_F32_X3`) on every platform. On GPU (A100, jax 0.10.2) this measurably changes a Held–Suarez T63L47 hybrid aquaplanet run with `spmd_mesh=(1,1,1)`. Global-mean ps is 0.1 hPa low from day 1, and the 40–60°N ps spin-up is ~4 hPa too weak by day 10 (1000.4 vs 1004.6 hPa). Setting only `FAST_TRANSFORM_DOT_ALGORITHM = Precision.HIGHEST` restores the single-device reference to ≤0.1 hPa.

This is *not* the constant-folding miscompile above: disabling XLA `constant_folding` leaves it unchanged. On CPU, X3 gives results identical to HIGHEST. We haven't isolated whether it is X3's inherent precision on GPU or another lowering issue. In jcm we now use HIGHEST for both defaults (~+2 % per step single-device, ~+8 % SPMD on a dycore-only case): climate-analytics-lab/jax-gcm#892.
